### PR TITLE
homework: support creating a home for a specified partition

### PR DESCRIPTION
**Pull request by pimzero**
This commit enable homework-luks to create a user home partition on an existing partition. This allows us to have multiple user home partitions on a single disk, or share user home partitions with the system partitions.

This commit address #15273 for "vanilla" partitions Partitions on LVM have not been tested.

This works by getting the parent block device of the selected partition, and if it has a GPT partition table, edit the partition table entry and format the partition as a per-user home partition.

Testing:

```
$ sudo mkosi --force build
[...]
$ truncate -s 16G ./mkosi.output/arch~rolling/image.raw && sfdisk -a ./mkosi.output/arch~rolling/image.raw <<EOF
label: gpt
device: /dev/sda
unit: sectors
sector-size: 512

/dev/sda3 : size=8G, uuid=$(systemd-id128 new -u), name=test
EOF
$ sudo mkosi qemu
[...]
[root@archlinux ~]# NEWPASSWORD=APassword111 homectl create user1 --storage=luks --image-path=/dev/sda3 --fs-type=ext4 # Testing with ext4 because of #22255
[root@archlinux ~]# homectl activate user1
[root@archlinux ~]# su user1
[user1@archlinux root]$ cd
[user1@archlinux ~]$ lsblk
NAME           MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINTS
sda              8:0    0   16G  0 disk
├─sda1           8:1    0  256M  0 part
├─sda2           8:2    0    3G  0 part  /
└─sda3           8:3    0    8G  0 part
  └─home-user1 254:0    0    8G  0 crypt /home/user1
```

I am not sure if this the interface that was envisioned by systemd team, but it seems to be the behaviour expected by the people seeing the issue #15273 , as well as what I wanted `homectl` to do. Please advise if a different behaviour would make sense. If this behaviour is indeed what we want, I will try to add tests next.

I don't have any experience with LVM, while I understand it could be meaningful to use it in this context, I didn't spend time trying to test with it. Should I try testing a setup with user home partitions on LVM ? Is there any other particular corner-cases I should think about ?

### CODE REVIEW

**Comment by bluca** [comment]
Please add a testcase for this in `test/units/testsuite-46.sh`

File: src/home/homework-luks.c [hunk 1]
```diff
@@ -1826,6 +1826,67 @@ static int luks_format(
...
+        assert(fdisk_partition_has_size(q));
+        size = fdisk_partition_get_size(q);
+        if (size > UINT64_MAX / 512U)
+                return log_error_errno(SYNTHETIC_ERRNO(ERANGE), "Partition size too large.");
+
+        *ret_offset = offset * 512U;
+        *ret_size = size * 512U;
+
+        return 0;
+}
```
Comment (poettering) [inline src/home/homework-luks.c:1898]: any chance you can split this into a separate preparation commit? i.e. that just splits out these two functions from make_partition_table() without adding the new logic just yet?

File: src/home/homework-luks.c [hunk 2]
```diff
@@ -1930,23 +1974,69 @@ static int make_partition_table(
...
+        c = fdisk_new_context();
+        if (!c)
+                return log_oom();
+
+        if (asprintf(&path, "/proc/self/fd/%i", fd) < 0)
+                return log_oom();
+
+        r = fdisk_assign_device(c, path, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open device: %m");
```
Comment (poettering) [inline src/home/homework-luks.c:2006]: these three operations here are the same as for make_partition_table(), and as I see now we have the same code elsewhere too, so maybe let's turn this into a proper helper:

→ #25503

can you rebase on top, and replace these three lines?

File: src/home/homework-luks.c [hunk 3]
```diff
@@ -1826,6 +1826,67 @@ static int luks_format(
         return 0;
 }
 
+static int make_partition(
+                const char *label,
+                sd_id128_t uuid,
+                struct fdisk_partition *p) {
```
Comment (poettering) [inline src/home/homework-luks.c]: hmm, this executes operations on `p`, hence please make that the first argument

File: src/home/homework-luks.c [hunk 4]
```diff
@@ -1930,23 +1974,69 @@ static int make_partition_table(
...
+
+        r = fdisk_write_disklabel(c);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write disk label: %m");
+
+        r = query_partition_offset_and_size(c, partno, &partition_offset, &partition_size);
+        if (r < 0)
+                return r;
+
+        r = sd_id128_from_string(fdisk_partition_get_uuid(p), ret_disk_uuid);
```
Comment (poettering) [inline src/home/homework-luks.c]: can you add an ASSERT_PTR() around the first arg? we know it must be non-NULL, because we just set it earlier, but let's make this clear here, given this was done in some other function

File: src/home/homework-luks.c [hunk 5]
```diff
@@ -2287,15 +2405,28 @@ int home_create_luks(
...
+                                setup->image_fd,
+                                user_record_user_name_and_realm(h),
+                                partno,
+                                partition_uuid,
+                                &partition_offset,
+                                &partition_size,
+                                &disk_uuid);
+                if (r < 0)
+                        return r;
+        }
```
Comment (poettering) [inline src/home/homework-luks.c]: you can move if `if (r < 0) return r` check outside of the outer if block, after all both branches have the same check

File: src/home/homework-luks.c [hunk 6]
```diff
@@ -2416,7 +2547,8 @@ int home_create_luks(
         if (r < 0)
                 return r;
 
-        setup->loop = loop_device_unref(setup->loop);
+        if (setup->loop)
+                setup->loop = loop_device_unref(setup->loop);
```
Comment (poettering) [inline src/home/homework-luks.c]: all our destructors are written so that passing NULL to them is a NOP, hence this change is unnecessary.

(this takes inspiration from libc free() which is also happy with NULL)

File: src/home/homework-luks.c [hunk 7]
```diff
@@ -2212,8 +2304,34 @@ int home_create_luks(
...
-                        return log_error_errno(SYNTHETIC_ERRNO(ENOTBLK), "Operating on partitions is currently not supported, sorry. Please specify a top-level block device.");
+                } else {
+                        _cleanup_free_ char *parent_path = NULL, *partno_buf = NULL;
+
+                        is_partition = true;
+                        r = block_get_whole_disk(st.st_rdev, &dev_parent);
+                        if (r < 0)
+                                return r;
+
+                        r = read_one_line_file(sysfs, &partno_buf);
```
Comment (poettering) [inline src/home/homework-luks.c]: pleas replace the access() above, with a call to read_virtual_file(). there's not too much point in first doing access() and then opening it. if we need to open it anyway in some cases, I think we should just do that and handle ENOENT from open() nicely.

(and yes, please use read_virtual_file() for this)

File: src/home/homework-luks.c [hunk 8]
```diff
@@ -2212,8 +2304,34 @@ int home_create_luks(
...
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to read partition number: %m");
+
+                        r = safe_atoi(partno_buf, &partno);
```
Comment (poettering) [inline src/home/homework-luks.c]: so weirdly, libfdisk makes partition numbers size_t. because that might vary in size on different archs, we so far made it uint64_t, i.e. the largest size_t on all archs we support. Please stick to that, i.e. make "partno" of type "uint64_t", just to remain systematic.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/home/homework-luks.c (hunk 4 above)
Comment (pimzero) [reply src/home/homework-luks.c]: Done - Though I am not sure if you want it around `p` or `fdisk_partition_get_uuid(p)`

File: src/home/homework-luks.c (hunk 7 above)
Comment (pimzero) [reply src/home/homework-luks.c]: Done. I put the max length of uint64_t base 10 represention as max_size, I don't know if another value makes more sense (`SIZE_MAX` ?)

File: src/home/homework-luks.c (hunk 1 above)
Comment (pimzero) [reply src/home/homework-luks.c:1898]: Done

File: src/home/homework-luks.c (hunk 2 above)
Comment (pimzero) [reply src/home/homework-luks.c:2006]: Thanks ! I rebased (on main now that it's been merged) and now use this helper

**Comment by pimzero** [comment]
> Please add a testcase for this in `test/units/testsuite-46.sh`

I am currently unable to make those tests work on my machine (I am probably missing a step to make it work) - I added something and will adapt depending on the results from the CI

File: src/home/homework-luks.c [hunk 9]
```diff
@@ -2235,11 +2287,37 @@ int home_create_luks(
 
                 if (asprintf(&sysfs, "/sys/dev/block/" DEVNUM_FORMAT_STR "/partition", DEVNUM_FORMAT_VAL(st.st_rdev)) < 0)
                         return log_oom();
-                if (access(sysfs, F_OK) < 0) {
+
+                r = read_virtual_file(sysfs, /* ⌈log10(UINT64_MAX)⌉ + 1 = */ 21, &partno_buf, NULL);
```
Comment (poettering) [inline src/home/homework-luks.c]: use `DECIMAL_STR_MAX(uint64_t) + 1` for the size

File: src/home/homework-luks.c [hunk 10]
```diff
@@ -2235,11 +2287,37 @@ int home_create_luks(
...
+                if (r < 0) {
                         if (errno != ENOENT)
                                 return log_error_errno(errno, "Failed to check whether %s exists: %m", sysfs);
-                } else
-                        return log_error_errno(SYNTHETIC_ERRNO(ENOTBLK), "Operating on partitions is currently not supported, sorry. Please specify a top-level block device.");
+                } else {
+                        _cleanup_free_ char *parent_path = NULL;
+
+                        partno_buf[r] = 0;
```
Comment (poettering) [inline src/home/homework-luks.c]: unnecessary, read_virtual_file() implicitly adds a NUL byte to the end of the returned buffer

File: src/home/homework-luks.c [hunk 11]
```diff
@@ -2313,13 +2391,24 @@ int home_create_luks(
...
-                        &disk_uuid);
+        if (!is_partition) {
+                r = make_partition_table(
+                                setup->image_fd,
+                                user_record_user_name_and_realm(h),
+                                partition_uuid,
+                                &partition_offset,
+                                &partition_size,
+                                &disk_uuid);
+        } else {
```
Comment (poettering) [inline src/home/homework-luks.c]: no {} around single line if blocks please, as per coding style

File: src/home/homework-luks.c [hunk 12]
```diff
@@ -2235,11 +2287,37 @@ int home_create_luks(
...
+                        partno--;
+
+                        r = device_path_make_canonical(S_IFBLK, dev_parent, &parent_path);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to build partition parent path: %m");
+
+                        setup->image_fd = open_image_file(h, parent_path, &st);
+                        if (setup->image_fd < 0)
+                                return setup->image_fd;
+                }
```
Comment (poettering) [inline src/home/homework-luks.c:2331]: the `host_size` calculations must be rearranged as well, i.e. the host_size should be at max the size of the partition, not of the whole block device

Comment (poettering) [reply src/home/homework-luks.c:2331]: and the discard logic below must take the partition offset/size into account as well, otherwise we might erase the whole disk by accident. people might end up being *very* pissed

**Review by poettering — CHANGES_REQUESTED** [review]
sorry for the late review!

**Comment by itspngu** [comment]
This would be a very useful feature - are you planning to rebase and address the requested changes @pimzero?

**Comment by dngray** [comment]
@poettering 

Might you have time to review this now that appears to be rebased?

**Comment by pimzero** [comment]
Hi @dngray , there are still comments that need to be addressed before asking for another review (that's why I put this MR as draft) - but unfortunately I did not have a lot of time to allocate for this. If anyone want to continue this feature, feel free. Apart from that, I can only apologize for taking reviewer time but not going to completion.

**Comment by dngray** [comment]
> there are still comments that need to be addressed before asking for another review

My bad there, I saw *outdated* and assumed they had been fixed. Unfortunately I don't really understand how this works, in order to complete it.
